ipld/amt/src/root.rs

```diff
@@ -1,62 +1,128 @@
 // Copyright 2019-2022 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
+use std::marker::PhantomData;
+
 use serde::de::{self, Deserialize};
 use serde::ser::{self, Serialize};
 
 use crate::node::CollapsedNode;
-use crate::{init_sized_vec, Node};
+use crate::{init_sized_vec, Node, DEFAULT_BIT_WIDTH};
+
+pub(crate) mod version {
+    #[derive(PartialEq, Debug)]
+    pub struct V0;
+    #[derive(PartialEq, Debug)]
+    pub struct V3;
+
+    pub trait Version {
+        const NUMBER: usize;
+    }
+
+    impl Version for V0 {
+        const NUMBER: usize = 0;
+    }
+
+    impl Version for V3 {
+        const NUMBER: usize = 3;
+    }
+}
 
 /// Root of an AMT vector, can be serialized and keeps track of height and count
+pub(super) type Root<V> = RootImpl<V, self::version::V3>;
+/// Legacy AMT v0, used to read block headers.
+pub(super) type Rootv0<V> = RootImpl<V, self::version::V0>;
+
 #[derive(PartialEq, Debug)]
-pub(super) struct Root<V> {
+pub(crate) struct RootImpl<V, Ver> {
     pub bit_width: u32,
     pub height: u32,
     pub count: u64,
     pub node: Node<V>,
+    ver: PhantomData<Ver>,
 }
 
-impl<V> Root<V> {
-    pub(super) fn new(bit_width: u32) -> Self {
+impl<V, Ver> RootImpl<V, Ver> {
+    pub(super) fn new_with_bit_width(bit_width: u32) -> Self {
         Self {
             bit_width,
             count: 0,
             height: 0,
             node: Node::Leaf {
                 vals: init_sized_vec(bit_width),
             },
+            ver: PhantomData,
+        }
+    }
+}
+
+impl<V> RootImpl<V, self::version::V0> {
+    pub(crate) fn new() -> Rootv0<V> {
```

Comment:
Let's make the visibility of `new` and `new_with_bit_width` match (just pick the same one, either one).

Reply:
Sure, using `pub(crate)`